### PR TITLE
Release File system feature for ESP32 (TC is passed)

### PR DESCRIPTION
**Pull request by ChinaTizenRT**
This PR aims to enable smartfs fileystem and mtd config partion,
and pass filesytem utc. (filesystem_tc), below it's main modification.

1) Enable filesystem  testcase config.
2) Add mtd config partion.
3) Fix mtd config partion  "GETCONFIG" crash issue.
4) Fix mtd config partion  "SETCONFIG" crash issue.

### CODE REVIEW

**Comment by ChinaTizenRT** [comment]
posting  test case log.
===============================
TASH>>filesystem_tc
TASH>>
########## FileSystem TC Start ##########

[tc_fs_vfs_umount] PASS 

[tc_fs_vfs_mount] PASS 

[tc_fs_vfs_open] PASS 

[tc_fs_vfs_write] PASS 

[tc_fs_vfs_read] PASS 

[tc_fs_vfs_close] PASS 

[tc_fs_vfs_dup] PASS 

[tc_fs_vfs_dup2] PASS 

[tc_fs_vfs_fsync] PASS 

[tc_fs_vfs_lseek] PASS 

[tc_fs_vfs_pwrite] PASS 

[tc_fs_vfs_pread] PASS 

[tc_fs_vfs_mkdir] PASS 

[tc_fs_vfs_opendir] PASS 

[tc_fs_vfs_readdir] PASS 

[tc_fs_vfs_rewinddir] PASS 

[tc_fs_vfs_seekdir] PASS 

[tc_fs_vfs_closedir] PASS 

[tc_fs_libc_dirent_readdir_r] PASS 

[tc_fs_libc_dirent_telldir] PASS 

[tc_fs_vfs_rmdir] PASS 

[tc_fs_vfs_unlink] PASS 

[tc_fs_vfs_stat] PASS 

[tc_fs_vfs_statfs] PASS 

[tc_fs_vfs_fstat] PASS 

[tc_fs_vfs_fstatfs] PASS 

[tc_fs_vfs_sendfile] PASS 

[tc_fs_vfs_fcntl] PASS 

[tc_fs_vfs_fdopen] PASS 

[tc_fs_vfs_rename] PASS 

[tc_fs_vfs_ioctl] PASS 
/proc_test/0/stat: 
0 0 0 2 2 1024 -1 -1 -1 Idle Task
/proc_test/0/status: 
Name:       Idle Task
Type:       Kernel thread
State:      Ready
Priority:   0 (0)
Scheduler:  SCHED_FIFO
SigMask:    00000000
/proc_test/0/cmdline: 
Idle Task
/proc_test/0/stack: 
StackBase:  0x0
StackSize:  1024
/proc_test/0/group/status: 
Main task:  0
Flags:      0x01
Members:    1
/proc_test/0/group/fd: 
FD  POS      OFLAGS
  0        0 0003
  1        0 0003
  2        0 0003
SD  RF TYP FLAGS
/proc_test/1/stat: 
1 0 224 34 6 2108 -1 -1 -1 hpwork
/proc_test/1/status: 
Name:       hpwork
Type:       Kernel thread
State:      Signal wait
Priority:   224 (224)
Scheduler:  SCHED_RR
SigMask:    00000000
/proc_test/1/cmdline: 
hpwork
/proc_test/1/stack: 
StackBase:  0x3ffe1848
StackSize:  2108
/proc_test/1/group/status: 
Main task:  1
Flags:      0x00
Members:    1
/proc_test/1/group/fd: 
FD  POS      OFLAGS
  0        0 0003
  1        0 0003
  2        0 0003
SD  RF TYP FLAGS
/proc_test/2/stat: 
2 0 176 34 6 2108 -1 -1 -1 lpwork
/proc_test/2/status: 
Name:       lpwork
Type:       Kernel thread
State:      Signal wait
Priority:   176 (176)
Scheduler:  SCHED_RR
SigMask:    00000000
/proc_test/2/cmdline: 
lpwork
/proc_test/2/stack: 
StackBase:  0x3ffe25d8
StackSize:  2108
/proc_test/2/group/status: 
Main task:  2
Flags:      0x00
Members:    1
/proc_test/2/group/fd: 
FD  POS      OFLAGS
  0        0 0003
  1        0 0003
  2        0 0003
SD  RF TYP FLAGS
/proc_test/3/stat: 
3 0 200 34 5 4152 -1 -1 -1 LWIP_TCP/IP
/proc_test/3/status: 
Name:       LWIP_TCP/IP
Type:       Kernel thread
State:      Semaphore wait
Priority:   200 (200)
Scheduler:  SCHED_RR
SigMask:    00000000
/proc_test/3/cmdline: 
LWIP_TCP/IP
/proc_test/3/stack: 
StackBase:  0x3ffe5524
StackSize:  4152
/proc_test/3/group/status: 
Main task:  3
Flags:      0x00
Members:    1
/proc_test/3/group/fd: 
FD  POS      OFLAGS
  0        0 0003
  1        0 0003
  2        0 0003
SD  RF TYP FLAGS
/proc_test/5/stat: 
5 4 125 32 5 4156 -1 -1 -1 tash
/proc_test/5/status: 
Name:       tash
Type:       Task
State:      Semaphore wait
Priority:   125 (125)
Scheduler:  SCHED_RR
SigMask:    00000000
/proc_test/5/cmdline: 
tash
/proc_test/5/stack: 
StackBase:  0x3ffe7848
StackSize:  4156
/proc_test/5/group/status: 
Main task:  5
Flags:      0x00
Members:    1
/proc_test/5/group/fd: 
FD  POS      OFLAGS
  0        0 0003
  1        0 0003
  2        0 0003
  3        0 0003
SD  RF TYP FLAGS
/proc_test/8/stat: 
8 5 100 32 3 4148 -1 -1 -1 filesystem_tc
/proc_test/8/status: 
Name:       filesystem_tc
Type:       Task
State:      Running
Priority:   100 (100)
Scheduler:  SCHED_RR
SigMask:    00000000
/proc_test/8/cmdline: 
filesystem_tc
/proc_test/8/stack: 
StackBase:  0x3ffe88a0
StackSize:  4148
/proc_test/8/group/status: 
Main task:  8
Flags:      0x00
Members:    1
/proc_test/8/group/fd: 
FD  POS      OFLAGS
  0        0 0003
  1        0 0003
  2        0 0003
  3        0 0003
  4        0 0001
SD  RF TYP FLAGS
/proc_test/fs/smartfs: 

/proc_test/mtd: 

/proc_test/partitions: 
Name        Start    Size   MTD
user            0     192   (null)

/proc_test/uptime: 
824633720915.21
/proc_test/version: 
Version: 2.0
Commit Hash: e15110d101aeb9bfbd6c8f4072cde4b1d3be2332
Build User: nonsmart@nonsmart
Build Time: 2019-03-27 15:50:57
/proc_test/fs/smartfs/smart0p0/debuglevel: 

/proc_test/fs/smartfs/smart0p0/status: 
Total Sectors    768
Free Sectors     744
Released Sectors 17


[tc_fs_smartfs_procfs_main] PASS 

[tc_driver_mtd_procfs_ops] PASS 

[tc_fs_procfs_main] PASS 

[tc_driver_mtd_config_ops] PASS 

[tc_libc_stdio_dprintf] PASS 

[tc_libc_stdio_fdopen] PASS 

[tc_libc_stdio_fopen] PASS 

[tc_libc_stdio_fclose] PASS 

[tc_libc_stdio_fputs] PASS 

[tc_libc_stdio_fgets] PASS 

[tc_libc_stdio_fseek] PASS 

[tc_libc_stdio_ftell] PASS 
Make long size file....please wait 
finished!
Yesterday all my trouble seemed so far away. Now it looks as though they're here to stay. Oh, I believe in yesterday.Yesterday all my trouble see.

[tc_libc_stdio_feof] PASS 

[tc_libc_stdio_fprintf] PASS 

[tc_libc_stdio_fsetpos] PASS 

[tc_libc_stdio_fgetpos] PASS 

[tc_libc_stdio_fputc] PASS 

[tc_libc_stdio_fgetc] PASS 

[tc_libc_stdio_fwrite] PASS 

[tc_libc_stdio_fread] PASS 

[tc_libc_stdio_freopen] PASS 

[tc_libc_stdio_ferror] PASS 

[tc_libc_stdio_clearerr] PASS 
Enter text here : 

[tc_libc_stdio_gets] PASS 
Enter text here : 

[tc_libc_stdio_gets_s] PASS 

[tc_libc_stdio_fileno] PASS 

[tc_libc_stdio_lib_snoflush] PASS 

[tc_libc_stdio_lib_sprintf] PASS 

mktemp: test1HhiboI
mktemp: test12IhiboI

[tc_libc_stdio_mktemp] PASS 

mkstemp: /tmp/test1OhiboI
mkstemp: /tmp/test12HdPaAL
mkstemp: /tmp/test2mkstemp
mkstemp: /tmp/test3gMNhEmp
mkstemp: /tmp/test4OImhGbXX

[tc_libc_stdio_mkstemp] PASS 

[tc_libc_stdio_remove] PASS 

[tc_libc_stdio_setbuf] PASS 

[tc_libc_stdio_setvbuf] PASS 

[tc_fs_mqueue_ops] PASS 

[tc_libc_stdio_meminstream] PASS 

[tc_libc_stdio_memoutstream] PASS 

[tc_libc_stdio_memsistream] PASS 

[tc_libc_stdio_memsostream] PASS 

[tc_libc_stdio_nullinstream] PASS 

[tc_libc_stdio_nulloutstream] PASS 

[tc_libc_stdio_rawinstream] PASS 

[tc_libc_stdio_rawoutstream] PASS 

[tc_libc_stdio_rawsistream] PASS 

[tc_libc_stdio_rawsostream] PASS 

[tc_libc_stdio_sprintf] PASS 

[tc_libc_stdio_stdinstream] PASS 

[tc_libc_stdio_stdoutstream] PASS 

[tc_libc_stdio_stdsistream] PASS 

[tc_libc_stdio_stdsostream] PASS 

tempnam: /tmp/tempnJiFjkF
tempnam: /tmp/EDbikF

[tc_libc_stdio_tempnam] PASS 

tmpnam: /tmp/fkadkF
tmpnam: /tmp/fDJjkF

[tc_libc_stdio_tmpnam] PASS 

[tc_libc_stdio_ungetc] PASS 

[tc_libc_stdio_zeroinstream] PASS 

[itc_fs_vfs_umount_n_twice] PASS 

[itc_fs_vfs_mount_p_read_mode] PASS 

[itc_fs_vfs_mount_n_twice] PASS 

[itc_fs_vfs_open_n] PASS 

[itc_fs_vfs_close_n] PASS 

[itc_fs_vfs_read_p_empty_file] PASS 

[itc_fs_vfs_write_read_p] PASS 

[itc_fs_vfs_fsync_p] PASS 

[itc_fs_vfs_lseek_n] PASS 

[itc_fs_vfs_lseek_n_closed_fd] PASS 

[itc_fs_vfs_lseek_p] PASS 

[itc_fs_vfs_pwrite_pread_p] PASS 

[itc_fs_vfs_mkdir_n] PASS 

[itc_fs_vfs_rmdir_n] PASS 

[itc_fs_vfs_mkdir_rmdir_n] PASS 

[itc_fs_vfs_opendir_n_invalid_path] PASS 

[itc_fs_vfs_opendir_n_file_path] PASS 

[itc_fs_vfs_readdir_rewinddir_p] PASS 

[itc_fs_vfs_seekdir_telldir_p] PASS 

[itc_fs_vfs_closedir_n] PASS 

[itc_libc_stdio_fdopen_p] PASS 

[itc_libc_stdio_fdopen_n] PASS 

[itc_libc_stdio_fopen_n_invalid_path] PASS 

[itc_libc_stdio_fopen_n_invalid_mode] PASS 

[itc_libc_stdio_fclose_n_null] PASS 

[itc_libc_stdio_fputs_n] PASS 

[itc_libc_stdio_fgets_n] PASS 

[itc_libc_stdio_fputs_fgets_p] PASS 

[itc_libc_stdio_fseek_n] PASS 

[itc_libc_stdio_ftell_n] PASS 

[itc_libc_stdio_fseek_ftell_p] PASS 

[itc_libc_stdio_fwrite_fread_p] PASS 
[tc_fs_smartfs_mksmartfs] PASS 

########## FileSystem TC End [PASS : 117, FAIL : 0] ##########

**Comment by seinfra** [comment]
Target : [01512063d50ce00c42908d5c623e404b5e6f0301] - Code Rule Check Result:
os/fs/driver/mtd/mtd_config.c:1164: ERROR: [SPC_M_KWD] space required before the open parenthesis '('

**Comment by seinfra** [comment]
Target : [01512063d50ce00c42908d5c623e404b5e6f0301] - Code Rule Check (C++) OK.

File: os/fs/driver/mtd/mtd_config.c [hunk 1]
```diff
@@ -1160,21 +1160,19 @@ static int mtdconfig_ioctl(FAR struct file *filep, int cmd, unsigned long arg)
 	FAR struct mtdconfig_struct_s *dev = inode->i_private;
 	FAR struct config_data_s *pdata;
 	int ret = -ENOSYS;
+	pdata = (FAR struct config_data_s *)arg;
+	if(!pdata) {
```
Comment (sunghan-chang) [inline os/fs/driver/mtd/mtd_config.c]: 1. It looks little strange.
Your code is `put data and then check validation`.
Logically, `put data after check validation` looks better.
If it is invalid, we don't need to re-assign to `pdata`.
2. Returning ENOSYS is not approprite. EINVAL is better.
```
if (!arg) {
    return -EINVAL;
}
pdata = (....)arg;
```

**Comment by sunghan-chang** [comment]
@ChinaTizenRT Thank you for PR and sharing result.
Could you check commit titles?
1. First commit is `build/configs: Fix config for ESP32 WIFI.` But 2nd commit has defconfig changes of wifi.
2. I don't know where you modify from 2nd commit title `apps/examples/testcase: Configure device name for ESP32.`. Please add `filesystem` on location.
All of testcases are in apps/examples so that `testcase/filesystem` is enough to say where it is I think.
3. To know what you modify, I should watch code change.  In my opinion, in history point view, what you do is better than what result is.
3rd commit can be `testcase/filesystem: change a way of buf variable definition to array in tc_driver_mtd_config_ops()`.
same as 4th commit.
4. There are violations of coding rules. It is difficult to say always by me and it is basic of sw development. Please check it by yourself first and then raise a PR to review.

File: os/fs/driver/mtd/mtd_config.c (hunk 1 above)
Comment (ChinaTizenRT) [reply os/fs/driver/mtd/mtd_config.c]: thanks for your suggestion, i agree and will fix it.

**Comment by chanijjani** [comment]
In title, Release Fileystem -> Release File system

**Comment by ChinaTizenRT** [comment]
as this PR is not well organized,  replaced by https://github.com/Samsung/TizenRT/pull/2857
close it here.

**Pull request closed without merging**